File: modules/mediapipe-pose/ios/MediaPipePoseModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import UIKit

/**
 * MediaPipePoseModule — iOS
 *
 * Status: PARTIAL — detectImage works, processVideo returns NOT_IMPLEMENTED.
 * Full AVAssetReader video pipeline deferred until Android architecture is proven.
 */
public class MediaPipePoseModule: Module {
  private var poseLandmarker: PoseLandmarker?
  private var currentVariant: String?

  public func definition() -> ModuleDefinition {
    Name("MediaPipePose")

    // ── isAvailable ──────────────────────────────────────────────────
    AsyncFunction("isAvailable") { (modelVariant: String) -> Bool in
      let modelName = self.modelName(for: modelVariant)
      return Bundle.main.path(forResource: modelName, ofType: "task") != nil
    }

    // ── detectImage ──────────────────────────────────────────────────
    AsyncFunction("detectImage") { (imageUri: String, modelVariant: String) -> [String: Any?] in
      if self.poseLandmarker == nil || self.currentVariant != modelVariant {
        try self.initializeLandmarker(modelVariant: modelVariant)
      }

      guard let landmarker = self.poseLandmarker else {
        throw NSError(domain: "MediaPipePose", code: 1,
          userInfo: [NSLocalizedDescriptionKey: "PoseLandmarker not initialized"])
      }

      guard let image = self.loadImage(from: imageUri) else {
        throw NSError(domain: "MediaPipePose", code: 2,
          userInfo: [NSLocalizedDescriptionKey: "Could not load image from URI: \(imageUri)"])
      }

      guard let mpImage = try? MPImage(uiImage: image) else {
        throw NSError(domain: "MediaPipePose", code: 3,
          userInfo: [NSLocalizedDescriptionKey: "Could not create MPImage"])
      }

      let startTime = CFAbsoluteTimeGetCurrent()
      let result = try landmarker.detect(image: mpImage)
      let inferenceDurationMs = (CFAbsoluteTimeGetCurrent() - startTime) * 1000.0

      return self.convertSingleResult(result: result, timestampMs: 0, inferenceDurationMs: inferenceDurationMs)
    }

    // ── processVideo — NOT IMPLEMENTED on iOS ────────────────────────
    AsyncFunction("processVideo") { (videoUri: String, modelVariant: String, targetFps: Double, maxFrames: Int) -> [String: Any] in
      throw NSError(domain: "MediaPipePose", code: 100,
        userInfo: [
          NSLocalizedDescriptionKey: "IOS_VIDEO_NOT_IMPLEMENTED: Video processing is not yet available on iOS. Use Android for R0 verification.",
          "code": "IOS_VIDEO_NOT_IMPLEMENTED"
        ])
    }

    // ── release ──────────────────────────────────────────────────────
    AsyncFunction("release") {
      self.poseLandmarker = nil
      self.currentVariant = nil
    }

    OnDestroy {
      self.poseLandmarker = nil
    }
  }

  // ── Private helpers ──────────────────────────────────────────────

  private func modelName(for variant: String) -> String {
    switch variant {
    case "full": return "pose_landmarker_full"
    case "heavy": return "pose_landmarker_heavy"
    default: return "pose_landmarker_lite"
    }
  }

  private func initializeLandmarker(modelVariant: String) throws {
    poseLandmarker = nil

    let name = modelName(for: modelVariant)
    guard let modelPath = Bundle.main.path(forResource: name, ofType: "task") else {
      throw NSError(domain: "MediaPipePose", code: 4,
        userInfo: [NSLocalizedDescriptionKey: "Model file not found: \(name).task. Run scripts/download-models.sh."])
    }

    let baseOptions = BaseOptions()
    baseOptions.modelAssetPath = modelPath

    let options = PoseLandmarkerOptions()
    options.baseOptions = baseOptions
    options.runningMode = .image
    options.numPoses = 1
    options.minPoseDetectionConfidence = 0.5
    options.minPosePresenceConfidence = 0.5
    options.minTrackingConfidence = 0.5

    poseLandmarker = try PoseLandmarker(options: options)
    currentVariant = modelVariant
  }

  private func convertSingleResult(result: PoseLandmarkerResult, timestampMs: Int, inferenceDurationMs: Double) -> [String: Any?] {
    guard !result.landmarks.isEmpty else {
      return [
        "timestampMs": timestampMs,
        "landmarks": [] as [[String: Double]],
        "worldLandmarks": [] as [[String: Double]],
        "inferenceDurationMs": inferenceDurationMs
      ]
    }

    let landmarks = result.landmarks[0]
    let worldLandmarks = result.worldLandmarks.isEmpty ? [] : result.worldLandmarks[0]

    let landmarkList: [[String: Double]] = landmarks.map { lm in
      [
        "x": Double(lm.x),
        "y": Double(lm.y),
        "z": Double(lm.z),
        "visibility": Double(lm.visibility?.doubleValue ?? 0),
        "presence": Double(lm.presence?.doubleValue ?? 0)
      ]
    }

    let worldLandmarkList: [[String: Double]] = worldLandmarks.map { lm in
      [
        "x": Double(lm.x),
        "y": Double(lm.y),
        "z": Double(lm.z),
        "visibility": Double(lm.visibility?.doubleValue ?? 0),
        "presence": Double(lm.presence?.doubleValue ?? 0)
      ]
    }

    return [
      "timestampMs": timestampMs,
      "landmarks": landmarkList,
      "worldLandmarks": worldLandmarkList,
      "inferenceDurationMs": inferenceDurationMs
    ]
  }

  private func loadImage(from uriString: String) -> UIImage? {
    if let url = URL(string: uriString), url.isFileURL {
      return UIImage(contentsOfFile: url.path)
    }
    return UIImage(contentsOfFile: uriString)
  }
}

File: modules/mediapipe-pose/android/src/main/java/expo/modules/mediapipepose/MediaPipePoseModule.kt
package expo.modules.mediapipepose

import android.content.Context
import android.graphics.BitmapFactory
import android.net.Uri
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import com.google.mediapipe.tasks.vision.core.RunningMode
import java.io.File
import java.io.FileNotFoundException

class MediaPipePoseModule : Module() {
  private var poseLandmarker: PoseLandmarker? = null
  private var currentVariant: String? = null

  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context is not available")

  override fun definition() = ModuleDefinition {
    Name("MediaPipePose")

    AsyncFunction("initialize") { modelVariant: String, promise: Promise ->
      try {
        initializeLandmarker(modelVariant)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_INIT", "Failed to initialize MediaPipe Pose Landmarker: ${e.message}", e)
      }
    }

    AsyncFunction("detectPose") { imageUri: String, modelVariant: String, promise: Promise ->
      try {
        // Auto-initialize if not done or variant changed
        if (poseLandmarker == null || currentVariant != modelVariant) {
          initializeLandmarker(modelVariant)
        }

        val landmarker = poseLandmarker
          ?: throw IllegalStateException("PoseLandmarker not initialized")

        // Load image from URI
        val bitmap = loadBitmapFromUri(imageUri)
          ?: throw FileNotFoundException("Could not load image from URI: $imageUri")

        val mpImage = BitmapImageBuilder(bitmap).build()

        // Run detection
        val result: PoseLandmarkerResult = landmarker.detect(mpImage)

        // Convert result to JS-compatible format
        val poses = mutableListOf<Map<String, Any>>()

        for (i in result.landmarks().indices) {
          val landmarks = result.landmarks()[i]
          val worldLandmarks = if (i < result.worldLandmarks().size) {
            result.worldLandmarks()[i]
          } else {
            emptyList()
          }

          val landmarkList = landmarks.map { lm ->
            mapOf(
              "x" to lm.x().toDouble(),
              "y" to lm.y().toDouble(),
              "z" to lm.z().toDouble(),
              "visibility" to (lm.visibility().orElse(0f)).toDouble(),
              "presence" to (lm.presence().orElse(0f)).toDouble()
            )
          }

          val worldLandmarkList = worldLandmarks.map { lm ->
            mapOf(
              "x" to lm.x().toDouble(),
              "y" to lm.y().toDouble(),
              "z" to lm.z().toDouble(),
              "visibility" to (lm.visibility().orElse(0f)).toDouble(),
              "presence" to (lm.presence().orElse(0f)).toDouble()
            )
          }

          poses.add(mapOf(
            "landmarks" to landmarkList,
            "worldLandmarks" to worldLandmarkList,
            "timestampMs" to System.currentTimeMillis()
          ))
        }

        promise.resolve(poses)
      } catch (e: Exception) {
        promise.reject("ERR_DETECT", "Pose detection failed: ${e.message}", e)
      }
    }

    AsyncFunction("release") { promise: Promise ->
      try {
        poseLandmarker?.close()
        poseLandmarker = null
        currentVariant = null
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_RELEASE", "Failed to release PoseLandmarker: ${e.message}", e)
      }
    }

    OnDestroy {
      poseLandmarker?.close()
      poseLandmarker = null
    }
  }

  private fun initializeLandmarker(modelVariant: String) {
    // Close existing instance if variant changed
    poseLandmarker?.close()

    val modelAssetName = when (modelVariant) {
      "lite" -> "pose_landmarker_lite.task"
      "full" -> "pose_landmarker_full.task"
      "heavy" -> "pose_landmarker_heavy.task"
      else -> "pose_landmarker_lite.task"
    }

    val baseOptions = BaseOptions.builder()
      .setModelAssetPath(modelAssetName)
      .build()

    val options = PoseLandmarker.PoseLandmarkerOptions.builder()
      .setBaseOptions(baseOptions)
      .setRunningMode(RunningMode.IMAGE)
      .setNumPoses(1)
      .setMinPoseDetectionConfidence(0.5f)
      .setMinPosePresenceConfidence(0.5f)
      .setMinTrackingConfidence(0.5f)
      .build()

    poseLandmarker = PoseLandmarker.createFromOptions(context, options)
    currentVariant = modelVariant
  }

  private fun loadBitmapFromUri(uriString: String): android.graphics.Bitmap? {
    return try {
      val uri = Uri.parse(uriString)
      if (uri.scheme == "file" || uri.scheme == null) {
        val path = uri.path ?: return null
        BitmapFactory.decodeFile(path)
      } else {
        val inputStream = context.contentResolver.openInputStream(uri)
        BitmapFactory.decodeStream(inputStream).also {
          inputStream?.close()
        }
      }
    } catch (e: Exception) {
      null
    }
  }
}
